build/ops: rpm: Package grafana dashboards

b-ranto opened the pull request:
This PR packages the grafana dashboards in a separate rpm package ceph-grafana-dashboards so that they could be later used by grafana server integrated with ceph-dashboard and they would get automatically updated on package upgrade. This also adds cmake support for installing the dashboards.

I have a test build repo with this branch, here:

https://1.chacra.ceph.com/r/ceph/wip-rpm-dashboards/a45d0ef99c9afd04e1eacd99de8710c6419b8318/centos/7/flavors/notcmalloc/

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

LenzGr commented on ceph.spec.in:
```diff
@@ -908,6 +908,17 @@ depending on python-ceph should be fixed to depend on python-rados,
+%endif
+%description grafana-dashboards
+This package provides a set of grafana dashboards for monitoring of
+ceph clusters. The dashboards require a Prometheus server setup with
```
Nit: "Ceph", not "ceph" maybe?

smithfarm commented on ceph.spec.in:
```diff
@@ -908,6 +908,17 @@ depending on python-ceph should be fixed to depend on python-rados,
 %endif
 
+%package grafana-dashboards
+Summary:	The set of grafana dashboards for monitoring purposes
```
While we're at it, could we have capital G in proper noun Grafana, please?

tchaikov replied:
since 4.6.0 RPM, `BuildArch: noarch` is supported, shall we add it for `ceph-grafana-dashboards` as it architecture independent configurations and presumably we don't need to support ancient rpm based distros .

smithfarm commented on ceph.spec.in:
```diff
@@ -908,6 +908,17 @@ depending on python-ceph should be fixed to depend on python-rados,
+Group:		System/Filesystems
+%endif
+%description grafana-dashboards
+This package provides a set of grafana dashboards for monitoring of
```
Grafana with capital G

smithfarm commented on ceph.spec.in:
```diff
@@ -908,6 +908,17 @@ depending on python-ceph should be fixed to depend on python-rados,
+%description grafana-dashboards
+This package provides a set of grafana dashboards for monitoring of
+ceph clusters. The dashboards require a Prometheus server setup with
+ceph-mgr prometheus collector and node_exporter collectors. These
```
Prometheus with capital P

b-ranto replied:
Here, I was referring to the prometheus collector/exporter ceph-mgr module (just like the node_exporter module/package). I suppose I should either capitalize both (Prometheus and Node Exporter) or keep them as they are?

smithfarm replied:
@b-ranto Ah, in that case, then drop the first instance of the word "collector" (since you have "collectors" at the end of the sentence). So it would read like this:

"The dashboards require a Prometheus server setup with the ceph-mgr "prometheus" and "node_exporter" collectors."

?

smithfarm replied:
(I like to put non-standard usage in double-quotes to call the reader's attention to the fact that it's a special case.)

tchaikov commented on ceph.spec.in:
```diff
@@ -908,6 +908,17 @@ depending on python-ceph should be fixed to depend on python-rados,
+%if 0%{?suse_version}
+Group:		System/Filesystems
+%endif
+%description grafana-dashboards
```
i wonder if grafana is a dependency of `ceph-grafana-dashboards`, shall we list it as a `Requires:` ?

jan--f replied:
This is a matter of opinion. The dashboards will sit there just fine without Grafana, but are useless. Some distros don't like packages like this fr this very reason.

I think its fine to leave as is, but have no strong feeling either way.

tchaikov replied:
i see... i thought grafana was packaged for well known distros. but turns out it's not =(

do you think is an option to  "Suggests: pcp-webapp-grafana" or "Suggest: grafana"? so dependency solver will ignore it and show them as hints?

jan--f replied:
Yes that certainly makes sense.

b-ranto replied:
I was thinking about requiring grafana, here. The thing is that grafana is likely to be run as a container (with /etc/grafana bind mounted), it wouldn't make much sense for the package to require it in that case.

I was thinking about doing something like `Suggests` but afaik, support for that is not very wide-spread, yet. :-/

tchaikov commented on ceph.spec.in:
```diff
@@ -1945,5 +1957,9 @@ exit 0
 
+%files grafana-dashboards
+%attr(0755,root,root) %dir %{_sysconfdir}/grafana/dashboards/ceph-dashboard
+%config %{_sysconfdir}/grafana/dashboards/ceph-dashboard/*
```
might want package `monitoring/grafana/dashboards`. it's difficult to figure out the dependency or how to use these settings without further pointers and instructions.

tchaikov commented on monitoring/grafana/dashboards/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+set(CMAKE_INSTALL_GRAFANA_DASHBOARDS "${CMAKE_INSTALL_SYSCONFDIR}/grafana/dashboards/ceph-dashboard"
```
i think `CMAKE_INSTALL_*` paths are set and/or use by cmake and its module(s), in this case, it's not a well known cmake variables out side of ceph, so i guess we could just use `CEPH_GRAFANA_DASHBOARDS_DIR` or `CEPH_GRAFANA_DASHBOARDS_PATH` here.

b-ranto commented:
I have updated the PR. The changes should hopefully address all of your concerns. I have made a couple more changes though:

- I have re-worded the description text; hopefully it is more clear and uses the capitalization properly, now
- I have added both README files to the `%doc` section for the package

b-ranto commented:
@smithfarm Ah, sorry, I have checked the terminal just now and I forgot to add `--force`, it should be fixed, now.

LenzGr commented:
@smithfarm or @tchaikov - would you mind merging this one? Thanks!

smithfarm commented:
Follow-on fix: https://github.com/ceph/ceph/pull/25347
